File: domain/src/test/java/co/joebirch/bastion/domain/interactor/GetTournamentsTest.java
package co.joebirch.bastion.domain.interactor;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.Collections;
import java.util.List;

import co.joebirch.bastion.domain.executor.PostExecutionThread;
import co.joebirch.bastion.domain.executor.ThreadExecutor;
import co.joebirch.bastion.domain.model.Tournament;
import co.joebirch.bastion.domain.repository.TournamentsRepository;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

import static org.hamcrest.CoreMatchers.instanceOf;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GetTournamentsTest {

    private GetTournaments getTournaments;

    @Mock private ThreadExecutor mockThreadExecutor;
    @Mock private PostExecutionThread mockPostExecutionThread;
    @Mock private TournamentsRepository mockTournamentsRepository;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        getTournaments = new GetTournaments(mockTournamentsRepository,
                mockThreadExecutor, mockPostExecutionThread);
    }

    @Test
    public void testGetTournamentsUseCaseObservableSucceeds() {
        getTournaments.buildUseCaseObservable(null);

        verify(mockTournamentsRepository).getTournaments();
        verifyNoMoreInteractions(mockTournamentsRepository);
        verifyZeroInteractions(mockThreadExecutor);
        verifyZeroInteractions(mockPostExecutionThread);
    }

    @Test
    public void testGetTournamentsUsecaseIsCalledOnce() {
        stubTournamentsRepositoryGetDisciplines();
        stubPostExecutionThreadGetScheduler();
        getTournaments.execute(null);

        verify(mockTournamentsRepository).getTournaments();
    }

    @Test
    public void testGetTournamentsUseCaseImplementsUseCase() {
        assertThat(getTournaments, instanceOf(UseCase.class));
    }

    private void stubTournamentsRepositoryGetDisciplines() {
        when(mockTournamentsRepository.getTournaments())
                .thenReturn(createDisciplineObservable());
    }

    private void stubPostExecutionThreadGetScheduler() {
        when(mockPostExecutionThread.getScheduler())
                .thenReturn(Schedulers.trampoline());
    }

    private Single<List<Tournament>> createDisciplineObservable() {
        return Single.just(Collections.singletonList(new Tournament()));
    }

}

File: domain/src/test/java/co/joebirch/bastion/domain/interactor/GetTournamentsTest.kt
package co.joebirch.bastion.domain.interactor

import co.joebirch.bastion.domain.model.Tournament
import co.joebirch.bastion.domain.util.TestDataFactory
import org.junit.Before
import org.junit.Test

class GetTournamentsTest {

    private lateinit var getTournaments: co.joebirch.bastion.domain.interactor.GetTournaments

    private var mockThreadExecutor: co.joebirch.bastion.domain.executor.ThreadExecutor = com.nhaarman.mockito_kotlin.mock()
    private var mockPostExecutionThread: co.joebirch.bastion.domain.executor.PostExecutionThread = com.nhaarman.mockito_kotlin.mock()
    private var mockTournamentsRepository: co.joebirch.bastion.domain.repository.TournamentsRepository = com.nhaarman.mockito_kotlin.mock()

    @Before
    fun setUp() {
        getTournaments = co.joebirch.bastion.domain.interactor.GetTournaments(mockTournamentsRepository,
                mockPostExecutionThread, mockThreadExecutor)
    }

    @Test
    fun testGetTournamentsUseCaseObservableSucceeds() {
        getTournaments.buildUseCaseObservable(null)

        com.nhaarman.mockito_kotlin.verify(mockTournamentsRepository).getTournaments()
        com.nhaarman.mockito_kotlin.verifyNoMoreInteractions(mockTournamentsRepository)
        com.nhaarman.mockito_kotlin.verifyZeroInteractions(mockThreadExecutor)
        com.nhaarman.mockito_kotlin.verifyZeroInteractions(mockPostExecutionThread)
    }

    @Test
    fun testGetTournamentsUsecaseIsCalledOnce() {
        stubTournamentsRepositoryGetDisciplines()
        stubPostExecutionThreadGetScheduler()
        getTournaments.execute(null)

        com.nhaarman.mockito_kotlin.verify(mockTournamentsRepository).getTournaments()
    }

    private fun stubTournamentsRepositoryGetDisciplines() {
        com.nhaarman.mockito_kotlin.whenever(mockTournamentsRepository.getTournaments())
                .thenReturn(createDisciplineObservable())
    }

    private fun stubPostExecutionThreadGetScheduler() {
        com.nhaarman.mockito_kotlin.whenever(mockPostExecutionThread.getScheduler())
                .thenReturn(io.reactivex.schedulers.Schedulers.trampoline())
    }

    private fun createDisciplineObservable(): io.reactivex.Single<List<Tournament>> {
        return io.reactivex.Single.just(listOf(co.joebirch.bastion.domain.util.TestDataFactory.Factory.makeTournament()))
    }

}